프로젝트 관리 앱 [STEP 3] Judy

Judy-999 opened the pull request:
# 프로젝트 관리 앱 Step 3
안녕하세요 라자냐 오랜만에 PR이네요 🤗
프로젝트 관리 앱 Step 3입니다!
기능만 구현하고 구조적으로 비효율적인 코드 같아 대대적으로 수정해보려다
라자냐의 리뷰를 받고 리팩토링 하는게 좋을 것 같아 이렇게 PR을 보냅니다!
<br>

## 코드 설명 👩🏻‍💻
Step 3에서 추가된 코드만 작성했습니다 🙂

**Network**
- `NetWorkMonitor`: 네트워크 연결을 감시하는 싱글톤 객체
<br>

**Database**
- `DatabaseManageable`: 데이터 관리를 추상화한 프로토콜
- `FirebaseService` : **Firebase**를 통한 데이터 로드 및 데이터 변환 기능을 제공하는 구조체
- `FirebaseManager`: **Firebase**에 저장, 수정, 삭제를 제공하는 싱글톤 객체
- `CoreDataManager` : **CoreData**에 저장, 수정, 삭제를 제공하는 싱글톤 객체
<br>

**View** 
- `HistoryTableViewCell` : 변경이력에 표현되는 테이블뷰 셀 [historyTitle, timestamp]
- `HistoryViewController` : 변경이력를 표현하는 뷰컨트롤러
<br>

**Util**
- `AlertManager` : 간단한 알림을 띄우는 기능을 제공하는 구조체

<br>


## 구현 화면 📱

**Firebase 저장**
![](https://i.imgur.com/kM35YxK.png)
<br>

| 인터넷 연결 확인 알림 |  변경 이력 에시 | 
| :--------: | :--------: |
|![](https://i.imgur.com/qktr9VM.png)|![](https://i.imgur.com/sODsZSN.png)|

<br>

## 고민한 및 궁금한 내용 🤔
### 1. Local DB - Remote DB의 역할 구분
로컬 저장소인 **CoreData**와 리모트 저장소인 **Firebase**의 역할을 고민했습니다. 

일반적으로는 **Firebase** 서버와 통신하며 동기화하다 네트워크 연결이 끊기면 **CoreData**로 전환하여 저장한 후 다시 연결이 되면 **Firebase**와 동기화한 후 **CoreData**에 저장된 데이터는 삭제되도록 했습니다. 

즉 모든 데이터는 **Firebase**에 저장되고 **CoreData**에는 인터넷 연결이 없는 동안 생성&변경된 데이터만 저장하도록 했습니다.

다만 생성과 수정의 경우는 잘 동작했지만 삭제의 경우 **CoreData** 상에서는 없는 데이터를 삭제하는 동작이 되어 원격과 동기화를 적용하지 못했습니다 😔

삭제한 데이터가 동기화되려면 로컬에서도 모든 데이터를 갖고 있게 하고, 인터넷 연결이 되면 리모트에 있는 데이터를 모두 삭제한 후 로컬 데이터로 덮어씌우는 방법을 생각했습니다. 하지만 로컬과 리모트 모두에 데이터를 저장하고 서로 덮어씌우는 방법이 너무 비효율적이란 생각이 들어 적용하진 않았습니다. 

제가 로컬과 원격 저장소의 역할을 너무 협소하게 분리한 것인지, 위와 같은 동기화 방법말고 다른 방법은 없는지 궁금합니다 😥
<br>

### 2. bind 코드 간소화
메인 화면을 표현하는 `ProjectManagerViewController`의 코드가 지나치게 긴 것 같아 할 일 상태별로 반복되던 작업을 `zip`과 `forEach`를 사용하여 간소화해주었습니다.

```swift
let workTables = [todoTableView, doingTableView, doneTableView]

zip(workTables, WorkState.allCases).forEach { workTable, state in
    viewModel.worksObservable
        .map {
            $0.filter { $0.state == state }
        }.bind(to: workTable.rx.items(cellIdentifier: // ..생략...
        }.disposed(by: disposeBag)
}
```
<br>

### 3. 비대한 뷰모델
현재 뷰모델은 `WorkViewModel` 하나만 존재하여 모든 데이터 조작을 관리하고 있습니다. 하나의 뷰모델이 많은 역할을 가지고 있고, 코드 역시 짧지 않아 분리가 필요할 것 같은데 어떤 기준으로 분리를 해야 할지 감이 잡히지 않습니다..!

별개로 이번 프로젝트에 적용할 만한 효율적인 역할 분담을 할 수 있는 구조나 패턴으로 추천해 주실 만한 것이 있을까요? 🧐

Judy-999 committed:
refactor: 수정 맟 추가 뷰컨에서 뷰모델을 init으로 받도록 변경

Judy-999 committed:
refactor: 반복되는 bind 코드 간소화

Judy-999 committed:
feat: Firebase 초기화

- GoogleService-info.plist 추가

Judy-999 committed:
feat: FirebaseService 구현

- loadData: Firestore에서 데이터를 받아오는 메서드
- convert: 스냅샷을 Work 타입으로 변환하는 메서드

Judy-999 committed:
feat: FirebaseManager를 생성해 FireStore의 CURD 구현

Judy-999 committed:
chore: GoogleService-info 재연결

Judy-999 committed:
feat: 뷰모델에서 Firestore에 저장 및 저장된 값을 가져오는 코드 구현

- fetch의 반환값을 Observable로 변경

Judy-999 committed:
feat: 뷰모델에서 Firestore에 저장 및 수정하는 코드 구현

Judy-999 committed:
feat: CoreData 모델인 ProjectManager 및 WorkEntity 생성

Judy-999 committed:
feat: CoreDataManager에 코어데이터 CURD 구현

- 뷰모델이 어떤 데이터 저장 방식을 사용할지 init으로 받도록 변경

Judy-999 committed:
feat: leftBarButton으로 History 추가

Judy-999 committed:
feat: Date의 convertToRegion에서 시간 스타일을 지정할 수 있게 변경

Judy-999 committed:
feat: WorkViewModel에 변경이력을 담는 histories 추가

Judy-999 committed:
feat: 변경이력을 popover로 띄울 HistoryViewController 구현

- 변경이력을 테이블뷰로 표시
- 변경이력과 시간을 표현하는 HistoryTableViewCell 구현

Judy-999 committed:
chore: View 그룹을 세부그룹으로 파일 정리

Judy-999 committed:
feat: 네트워크 연결 확인하는 싱글톤 객체 생성

Judy-999 committed:
style:  ProjectManagerViewController의 코드 구조 변경

- 지나친 함수 분리 병합

Judy-999 committed:
refactor: NetWorkMonitor에서 Observable을 가지도록 변경

Judy-999 committed:
feat: Network 연결 상태에 따라 원격-로컬을 변경

Judy-999 committed:
feat: 인터넷이 다시 연결되면 코어데이터의 데이터를 원격으로 백업 후 삭제

Judy-999 committed:
feat: 인터넷 연결이 끊기면 사용자에게 얼럿으로 알려주는 코드 작성

- 간단한 얼럿을 생성하는 AlertManager 구조체 생성

Judy-999 committed:
refactor: 상속하지 않는 클래스에 final 키워드 적용

- 할 일의 추가 및 수정하는 뷰를 Manage -> Detail 네이밍 변경
- 사용하지 않는 SampleData 삭제

Judy-999 committed:
refactor: 최신 변경이력이 상단에 추가되도록 수정

wonhee009 commented on ProjectManager/ProjectManager/View/Detail/WorkDetailViewController.swift:
```diff
@@ -62,44 +72,41 @@ final class ManageWorkViewController: UIViewController {
         switch viewMode {
         case .add:
-            guard let newWork = workManageView.createNewWork() else { return }
+            guard let newWork = workDetailView.createNewWork() else { return }
```
`createNewWork`를 `workDetailView`에서 하는 이유가 있을까요?
비즈니스 로직을 view가 하는거처럼 생각됩니다.

Judy-999 replied:
텍스트필드와 같은 UI 요소들을 `private`으로 하여 `workDetailView`에서만 접근할 수 있기 때문에 뷰에서 생성하도록 했습니다.

`workDetailView`를 사용하지 않고 `workDetailViewController`에서 직접 UI를 구현하여 뷰모델에서 `createWork` 메서드를 통해 생성하도록 변경했습니다!

wonhee009 replied:
`createWork` 로직이 비즈니스 로직이라고 생각이 드는데요. ViewModel에서 해당 작업 하는건 어떤가요?

Judy-999 replied:
아아 넵! 새로운 할 일을 생성하는 `createWork`는 ViewModel에 구현했습니다.
```swift
    func createWork(_ id: UUID, _ title: String, _ content: String, _ deadline: Date, _ state: WorkState) {
        let work = Work(id: id, title: title, content: content, deadline: deadline, state: state)
        
        newWork.onNext(work)
    }
```
`workDetailViewController`에선 UI 요소와 연결만 해주었습니다!

wonhee009 commented on ProjectManager/ProjectManager/View/Detail/WorkDetailViewController.swift:
```diff
@@ -62,44 +72,41 @@ final class ManageWorkViewController: UIViewController {
         case .edit:
             guard let work = work,
-                  let newWork = self.workManageView.createNewWork(id: work.id, state: work.state) else { return }
+                  let newWork = self.workDetailView.createNewWork(id: work.id, state: work.state) else { return }
```
`edit` 모드인 경우에도 새로운 work를 생성하는건가요?

Judy-999 replied:
넵! 
(제목, 내용, 날짜) 중 어떤 값이 변경될지 모르기도 하고, `Work`의 프로퍼티들을 `var`로 하는 것보다 새로운 `Work`를 생성해 변경하는 방법을 선택했습니다!

wonhee009 commented on ProjectManager/ProjectManager/View/Detail/WorkDetailViewController.swift:
```diff
@@ -62,44 +72,41 @@ final class ManageWorkViewController: UIViewController {
-        workManageView.changeEditMode(false)
+    func configureEditMode(with work: Work) {
+        workDetailView.configure(with: work)
+        workDetailView.changeEditMode(false)
```
`configureEditMode` 네이밍으로 미뤄봤을때 수정 모드일때 설정으로 보이는데요.
`changeEditMode(true)`가 되어야 하지 않나요?

Judy-999 replied:
**EditMode**일 때 처음엔 `false`로 설정하고 `Edit` 버튼을 누르면 `true`로 변경하여 수정할 수 있도록 했는데 
지금 보니 수정화면일 때 `false`로 설정해두는 로직이 어색한 것 같네요!

기본 설정을 수정 불가능하도록 하고

1. **AddMode**일 때
2. **EditMode**에서 `Edit` 버튼을 누를 때

위 두 경우에 수정 가능하게 변경하도록 했습니다.
오로지 수정 가능하게 변경하는 메서드로 역할이 변경되어 네이밍 역시 `switchEditable`로 수정했습니다!

wonhee009 commented on ProjectManager/ProjectManager/View/Main/ProjectManagerViewController.swift:
```diff
@@ -68,186 +69,122 @@ final class ProjectManagerViewController: UIViewController {
     @objc private func addWorkBarButtonTapped() {
-        let manageViewController = ManageWorkViewController()
-        manageViewController.configureAddMode(viewModel)
+        let manageViewController = WorkDetailViewController(viewModel: viewModel)
```
프로퍼티 네이밍과 VC 네이밍간의 관계가 보이면 좋을거 같습니다.

Judy-999 replied:
네이밍 변경하면서 발견하지 못했네요..! 수정했습니다.

wonhee009 commented on ProjectManager/ProjectManager/View/Main/ProjectManagerViewController.swift:
```diff
@@ -68,186 +69,122 @@ final class ProjectManagerViewController: UIViewController {
-            .asDriver(onErrorJustReturn: nil)
-            .drive(doingTitleView.countImageView.rx.image)
-            .disposed(by: disposeBag)
+        zip(workTables, WorkState.allCases).forEach { workTable, state in
```
`WorkState`의 case 순서와 workTables의 Element 순서를 맞춰줘야 제대로 동작할거 같은데요.
나중에 Case가 추가 되었을때 개발자의 작성 실수가 있을 수 있지 않을까요?
순서를 보장할 수 있는 보조적인 부분이 필요할거 같습니다.

Judy-999 replied:
`allCases` 대신에 배열을 사용하도록 변경했습니다!

```swift
let workStates = [WorkState.todo, WorkState.doing, WorkState.done]
```

wonhee009 replied:
`allCases` 대샌 배열을 사용해도 workTables, workStates의 순서를 맞춰야하는건 동일할거 같습니다.
`enum WorkState: CaseIterable, Int`를 사용해 rawValue를 부여하고 해당 rawValue를 이용해서 allCases를 정렬하고, workTables도 `WorkState`를 활용하는 방식으로 변경한다면 어떤 case가 추가되어도 항상 싱크 맞출 수 있을거 같은데 어떻게 생각하실까요?

Judy-999 replied:
```swift
enum WorkState: Int, CaseIterable {
    case todo = 0
    case doing 
    case done
    
    var title: String {
        switch self {
        case .todo:
            return "TODO"
        case .doing:
            return "DOING"
        case .done:
            return "DONE"
        }
    }
}
```
기존에 사용하던 `String` rawValue는 연산 프로퍼티로 대체하고 `Int`로 변경했습니다.
<br>

```swift
let workTables = [todoTableView, doingTableView, doneTableView].sorted(by: { $0.tag < $1.tag })
let headerViews = [toDoTitleView, doingTitleView, doneTitleView].sorted(by: { $0.tag < $1.tag })
let workStates = WorkState.allCases.sorted(by: { $0.rawValue < $1.rawValue })
```
이후 `WorkState`의 allCases를 사용할 때 rawValue로 정렬하고,
TableView와 TitleView 역시 초기화할 때 `WorkState`를 받아 `tag`를 rawValue 값으로 설정하여 마찬가지로 `WorkState`와 같은 순서로 정렬 가능하도록 했습니다.

라자냐가 설명해주신 내용이 이 방식이 맞을까요?!

wonhee009 commented on ProjectManager/ProjectManager/ViewModel/WorkViewModel.swift:
```diff
@@ -29,12 +57,22 @@ class WorkViewModel {
+        
+        histories
+            .take(1)
+            .observe(on: MainScheduler.instance)
```
여기서 `MainScheduler`가 필요한가요?
VM에서 `MainScheduler`를 잡고 있어 질문 드립니다.

Judy-999 replied:
필요하지 않습니다..
변경이력을 추가하는거라 UI와 관련돼있다고 착각해서 Main으로 해주었는데 다시 생각해보니 그럴 필요가 없는 코드라 제거했습니다!

wonhee009 commented on ProjectManager/ProjectManager/WorkEntity+CoreDataProperties.swift:
```diff
@@ -0,0 +1,28 @@
+
+    @NSManaged public var id: UUID
+    @NSManaged public var title: String
+    @NSManaged public var content: String
```
항상 non-Optional인가요?

Judy-999 replied:
할 일(`Work`)를 생성할 때 항상 모든 값이 존재하도록 하기 때문에 CoreData에 저장하거나 가져올 때도 non-Optional한 상태라 생각하여 `?`을 제거했습니다!

혹시 Entity에서 제가 모르는 Optional한 상태가 발생할 수도 있는 것일까요?

wonhee009 replied:
아뇨 제 생각에는 content가 없이도 할일이 만들어질 수 있지 않나 생각해서 질문 드렸습니다.

Judy-999 replied:
현재 `content`로 아무것도 입력하지 않으면 `""`인 상태로 저장됩니다. 
값이 없음은 아니고 빈 칸이란 값이 있어서 옵셔널로 표현하진 않았습니다.
내용이 없음을 빈 칸으로 표현한다고 생각했는데 옵셔널로 표현해야 할까요?

wonhee009 replied:
사용자가 실제로 입력한 부분은 아녀서 저는 옵셔널로 생각했습니다 :)

Judy-999 replied:
```swift
extension WorkEntity {
    @NSManaged public var content: String?
    @NSManaged public var deadline: Date
    @NSManaged public var id: UUID
    //...
}
```

사용자의 입력은 아니니 옵셔널로 표현하는 것이 의미에 부합한 것 같아 변경해주었습니다!
<br> 

```swift
if let content = $0.content {
    return Work(id: $0.id, title: $0.title, content: content, deadline: $0.deadline, state: state)
} //...
```
저장할 때 값이 `""`라면 `content`를 저장하지 않도록 변경하고 fetch할 때도 위 코드처럼 `nil`인지 검사하도록 변경했습니다.

wonhee009 commented on ProjectManager/ProjectManager/Util/AlertManager.swift:
```diff
@@ -0,0 +1,17 @@
+
+import UIKit
+
+struct AlertManager {
+    func showAlert(_ controller: UIViewController, title: String, message: String) {
+        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
+        
+        alert.addAction(UIAlertAction(title: "확인", style: .cancel))
+        controller.present(alert, animated: true)
+    }
+}
```
이후에 "확인/취소" 버튼이 있는 alert이 추가된다면 메서드가 추가되는 형태일까요?

Judy-999 replied:
현재는 그럴 생각이었습니다!
만약 필요한 alert 형태가 많아진다면 하나의 메서드로 구현할 수 있는 방법을 고민해보긴 해야겠네요..! 🧐

wonhee009 reviewed (changes requested):
수고하셨습니다 :)
코멘트 남겨두었으니 확인 부탁드립니다.

- DB 역할 구분: 현재 프로젝트에서 리모트 DB를 사용하는 이유가 뭘까요? 저는 해당 앱을 다른 디바이스에서도 동일하게 사용하기 위해 사용한다고 이해했는데요. (+ 오프라인) 따라서 저는 리모트와 로컬 DB끼리 동기화되는게 맞다고 생각했습니다.
- 비대한 뷰모델: 비즈니스 로직의 역할에 따라 나누긴 하는데요. 현재 구조상 분리가 어렵다면 굳이 분리하지 않아도 될거 같습니다. (대신 역할을 제대로 나눴는지는 다시 한 번 점검해봐야겠지만요.)

Judy-999 committed:
chore: 네이밍 통일성 부여

Judy-999 committed:
refactor: DetailView를 사용하지 않고 뷰컨트롤러에서 뷰를 구현하도록 변경

- 새로운 할 일을 뷰모델이서 생성하도록 수정

Judy-999 committed:
refactor: 할 일이 생성 또는 변경되면 fetch를 통해 변경사항 적용

- works에 일일이 변경사항을 반영하는 대신 db에서 fetch해온 데이터로 변경

Judy-999 committed:
refactor: 변경이력에 내용 추가하는 로직 변경

- HistoryType을 생성하고 변경 이력을 추가하는 메서드 구현

Judy-999 committed:
refactor: WorkState의 allCases 대신 배열을 생성해 사용

Judy-999 committed:
feat: DatabaseManager로 리모트와 로컬 동기화

Judy-999 commented:
리뷰 감사합니다 라자냐 😊
코멘트 기반으로 리팩토링 완료했습니다!
<br>

> DB 역할 구분: 현재 프로젝트에서 리모트 DB를 사용하는 이유가 뭘까요? 저는 해당 앱을 다른 디바이스에서도 동일하게 사용하기 위해 사용한다고 이해했는데요. (+ 오프라인) 따라서 저는 리모트와 로컬 DB끼리 동기화되는게 맞다고 생각했습니다.

`DatabaseManager`라는 타입을 만들어 로컬과 리모트에 데이터에 변경이 발생하면 로컬과 리모트 모두에게 반영되도록 변경했습니다.
<br>

> 비대한 뷰모델: 비즈니스 로직의 역할에 따라 나누긴 하는데요. 현재 구조상 분리가 어렵다면 굳이 분리하지 않아도 될거 같습니다. (대신 역할을 제대로 나눴는지는 다시 한 번 점검해봐야겠지만요.)

뷰모델은 분리해보려 고민하다가 많이 복잡한 로직은 아닌 것 같아 코드 간소화만 진행했습니다.

Judy-999 committed:
chore: 사용하지 않는 코드 제거

Judy-999 committed:
refactor: WorkState의 rawValue로 표시한 내용을 연산 프로퍼티로 변경

Judy-999 committed:
refactor: WorkState의 rawValue를 Int로 변경

- WorkState의 allCases를 사용할 때 rawValue로 정렬
- HeaderView와 WorkTableView의 init에서 WorkState rawValue로 tag 초기화
- WorkEntity의 state를 String에서 Int64로 변경

Judy-999 committed:
refactor: content를 입력하지 않을 경우 "" 대신 nil이 되도록 변경

Judy-999 committed:
refactor: 뷰모델의 init에서 사용할 DB를 입력받지 않도록 변경

- 특정 DB 대신에 DatabaseManager를 사용

Judy-999 committed:
docs: Step 3 README 작성

wonhee009 reviewed (approved):
수고하셨습니다 :)
